### ISSUE TITLE
[clang++] crash when compiling templated method in templated class

### ISSUE THREAD

**Issue by FerdinandSpitzschnueffler** [llvm/llvm-project#62331]
When compiling the source code below with the command `clang++ essence.cpp`, the compiler crashes with a segmentation fault.

## Affected Versions

clang-15, clang-16, clang-17

## Source code

```cxx
// file essence.cpp
template <typename>
struct SomeImpl {};

template <template <typename> class T1>
class A1 {
   public:
    template <typename T2>
    void b(T1<T2> = T1<T2>());
};

template <>
template <typename T2>
void A1<SomeImpl>::b(SomeImpl<T2>) {}

void bug1() {
    A1<SomeImpl> a;
    a.b<int>();
    //When providing an argument the compiler does not crash.
    //a.b<int>(SomeImpl<int>());
}

int main() {}
```

## Crash Report

```console
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /usr/lib/llvm-17/bin/clang -cc1 -triple x86_64-pc-linux-gnu -emit-obj -mrelax-all -disable-free -clear-ast-before-backend -disable-ll
vm-verifier -discard-value-names -main-file-name essence.cpp -mrelocation-model pic -pic-level 2 -pic-is-pie -mframe-pointer=all -fmath-errno -ffp-contract=on -
fno-rounding-math -mconstructor-aliases -funwind-tables=2 -target-cpu x86-64 -tune-cpu generic -debugger-tuning=gdb -fcoverage-compilation-dir=/home/user/Development/tmp -resource-dir /usr/lib/llvm-17/lib/clang/17 -internal-isystem /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12 -internal-i
system /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/x86_64-linux-gnu/c++/12 -internal-isystem /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../..
/include/c++/12/backward -internal-isystem /usr/lib/llvm-17/lib/clang/17/include -internal-isystem /usr/local/include -internal-isystem /usr/bin/../lib/gcc/x86_
64-linux-gnu/12/../../../../x86_64-linux-gnu/include -internal-externc-isystem /usr/include/x86_64-linux-gnu -internal-externc-isystem /include -internal-extern
c-isystem /usr/include -fdeprecated-macro -fdebug-compilation-dir=/home/user/Development/tmp -ferror-limit 19 -fgnuc-version=4.2.1 -fcxx-exceptions -f
exceptions -fcolor-diagnostics -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o /tmp/essence-6e6535.o -x c++ essence.cpp
1.	essence.cpp:18:14: current parser token ')'
2.	essence.cpp:16:13: parsing function body 'bug1'
3.	essence.cpp:16:13: in compound statement ('{}')
#0 0x00007fb09e422ac6 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/usr/lib/llvm-17/bin/../lib/libLLVM-17.so.1+0xc22ac6)
#1 0x00007fb09e420c60 llvm::sys::RunSignalHandlers() (/usr/lib/llvm-17/bin/../lib/libLLVM-17.so.1+0xc20c60)
#2 0x00007fb09e42315b (/usr/lib/llvm-17/bin/../lib/libLLVM-17.so.1+0xc2315b)
#3 0x00007fb09d042520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
clang++-17: error: unable to execute command: Segmentation fault (core dumped)
clang++-17: error: clang frontend command failed due to signal (use -v to see invocation)
Ubuntu clang version 17.0.0 (++20230417095441+43ac269bdd00-1~exp1~20230417215605.872)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
clang++-17: note: diagnostic msg:
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang++-17: note: diagnostic msg: /tmp/essence-99723d.cpp
clang++-17: note: diagnostic msg: /tmp/essence-99723d.sh
clang++-17: note: diagnostic msg:

********************
```

## Compile Command

```sh
# Crash reproducer for Ubuntu clang version 17.0.0 (++20230417095441+43ac269bdd00-1~exp1~20230417215605.872)
# Driver args: "--driver-mode=g++" "-c" "essence.cpp" "-emit-llvm" "-Xclang" "-disable-llvm-passes"
# Original command:  "/usr/lib/llvm-17/bin/clang" "-cc1" "-triple" "x86_64-pc-linux-gnu" "-emit-llvm-bc" "-emit-llvm-uselists" "-disable-free" "-clear-ast-before-backend" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "essence.cpp" "-mrelocation-model" "pic" "-pic-level" "2" "-pic-is-pie" "-mframe-pointer=all" "-fmath-errno" "-ffp-contract=on" "-fno-rounding-math" "-mconstructor-aliases" "-funwind-tables=2" "-target-cpu" "x86-64" "-tune-cpu" "generic" "-debugger-tuning=gdb" "-fcoverage-compilation-dir=/home/user/Development/tmp" "-resource-dir" "/usr/lib/llvm-17/lib/clang/17" "-internal-isystem" "/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12" "-internal-isystem" "/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/x86_64-linux-gnu/c++/12" "-internal-isystem" "/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/backward" "-internal-isystem" "/usr/lib/llvm-17/lib/clang/17/include" "-internal-isystem" "/usr/local/include" "-internal-isystem" "/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../x86_64-linux-gnu/include" "-internal-externc-isystem" "/usr/include/x86_64-linux-gnu" "-internal-externc-isystem" "/include" "-internal-externc-isystem" "/usr/include" "-fdeprecated-macro" "-fdebug-compilation-dir=/home/user/Development/tmp" "-ferror-limit" "19" "-fgnuc-version=4.2.1" "-fcxx-exceptions" "-fexceptions" "-fcolor-diagnostics" "-disable-llvm-passes" "-faddrsig" "-D__GCC_HAVE_DWARF2_CFI_ASM=1" "-o" "essence.bc" "-x" "c++" "essence.cpp"
 "/usr/lib/llvm-17/bin/clang" "-cc1" "-triple" "x86_64-pc-linux-gnu" "-emit-llvm-bc" "-emit-llvm-uselists" "-disable-free" "-clear-ast-before-backend" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "essence.cpp" "-mrelocation-model" "pic" "-pic-level" "2" "-pic-is-pie" "-mframe-pointer=all" "-fmath-errno" "-ffp-contract=on" "-fno-rounding-math" "-mconstructor-aliases" "-funwind-tables=2" "-target-cpu" "x86-64" "-tune-cpu" "generic" "-debugger-tuning=gdb" "-fcoverage-compilation-dir=/home/user/Development/tmp" "-fdeprecated-macro" "-fdebug-compilation-dir=/home/user/Development/tmp" "-ferror-limit" "19" "-fgnuc-version=4.2.1" "-fcxx-exceptions" "-fexceptions" "-fcolor-diagnostics" "-disable-llvm-passes" "-faddrsig" "-D__GCC_HAVE_DWARF2_CFI_ASM=1" "-x" "c++" "essence-5c4631.cpp"
```

**Comment by DimitryAndric** [comment]
This is a very old bug, already reported in #20832 (in 2014!). With the most recent clang I have here:

```
Assertion failed: (getKind() == Template && "Unexpected kind"), function getAsTemplate, file /home/dim/src/llvm/llvm-project/clang/include/clang/AST/TemplateBase.h, line 312.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: /home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang -cc1 -triple x86_64-- -S pr62331-1.cpp
1.      pr62331-1.cpp:18:14: current parser token ')'
2.      pr62331-1.cpp:16:13: parsing function body 'bug1'
3.      pr62331-1.cpp:16:13: in compound statement ('{}')
 #0 0x00000000042e6458 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x42e6458)
 #1 0x00000000042e4368 llvm::sys::RunSignalHandlers() (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x42e4368)
 #2 0x00000000042e6be0 SignalHandler(int) Signals.cpp:0:0
 #3 0x000000082a01ba60 handle_signal /usr/src/lib/libthr/thread/thr_sig.c:0:3
 #4 0x000000082a01b01f thr_sighandler /usr/src/lib/libthr/thread/thr_sig.c:247:1
 #5 0x00000008284658a3 ([vdso]+0x2d3)
 #6 0x000000082f1f092a __sys_thr_kill /usr/obj/usr/src/lib/libc/thr_kill.S:4:0
 #7 0x000000082f1699c4 __raise /usr/src/lib/libc/gen/raise.c:0:10
 #8 0x000000082f21a779 abort /usr/src/lib/libc/stdlib/abort.c:73:17
 #9 0x000000082f14c7a1 (/lib/libc.so.7+0x937a1)
#10 0x0000000006568878 (anonymous namespace)::TemplateInstantiator::TransformTemplateName(clang::CXXScopeSpec&, clang::TemplateName, clang::SourceLocation, clang::QualType, clang::NamedDecl*, bool) SemaTemplateInstantiate.cpp:0:0
#11 0x000000000658a507 clang::TreeTransform<(anonymous namespace)::TemplateInstantiator>::TransformTemplateSpecializationType(clang::TypeLocBuilder&, clang::TemplateSpecializationTypeLoc) SemaTemplateInstantiate.cpp:0:0
#12 0x000000000655ed08 clang::TreeTransform<(anonymous namespace)::TemplateInstantiator>::TransformType(clang::TypeLocBuilder&, clang::TypeLoc) SemaTemplateInstantiate.cpp:0:0
#13 0x00000000065879b0 clang::TreeTransform<(anonymous namespace)::TemplateInstantiator>::TransformElaboratedType(clang::TypeLocBuilder&, clang::ElaboratedTypeLoc) SemaTemplateInstantiate.cpp:0:0
#14 0x000000000655e926 clang::TreeTransform<(anonymous namespace)::TemplateInstantiator>::TransformType(clang::TypeLocBuilder&, clang::TypeLoc) SemaTemplateInstantiate.cpp:0:0
#15 0x000000000655e29b clang::TreeTransform<(anonymous namespace)::TemplateInstantiator>::TransformType(clang::TypeSourceInfo*) SemaTemplateInstantiate.cpp:0:0
#16 0x000000000656eb64 clang::TreeTransform<(anonymous namespace)::TemplateInstantiator>::TransformCXXUnresolvedConstructExpr(clang::CXXUnresolvedConstructExpr*) SemaTemplateInstantiate.cpp:0:0
#17 0x000000000658e3ee void llvm::function_ref<void ()>::callback_fn<clang::Sema::SubstDefaultArgument(clang::SourceLocation, clang::ParmVarDecl*, clang::MultiLevelTemplateArgumentList const&, bool)::$_1>(long) SemaTemplateInstantiate.cpp:0:0
#18 0x0000000005cdbf5e clang::Sema::runWithSufficientStackSpace(clang::SourceLocation, llvm::function_ref<void ()>) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x5cdbf5e)
#19 0x00000000065615cb clang::Sema::SubstDefaultArgument(clang::SourceLocation, clang::ParmVarDecl*, clang::MultiLevelTemplateArgumentList const&, bool) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x65615cb)
#20 0x00000000065b6f50 clang::Sema::InstantiateDefaultArgument(clang::SourceLocation, clang::FunctionDecl*, clang::ParmVarDecl*) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x65b6f50)
#21 0x0000000006044837 clang::Sema::BuildCXXDefaultArgExpr(clang::SourceLocation, clang::FunctionDecl*, clang::ParmVarDecl*, clang::Expr*) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x6044837)
#22 0x0000000006047540 clang::Sema::GatherArgumentsForCall(clang::SourceLocation, clang::FunctionDecl*, clang::FunctionProtoType const*, unsigned int, llvm::ArrayRef<clang::Expr*>, llvm::SmallVectorImpl<clang::Expr*>&, clang::Sema::VariadicCallType, bool, bool) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x6047540)
#23 0x0000000006046103 clang::Sema::ConvertArgumentsForCall(clang::CallExpr*, clang::Expr*, clang::FunctionDecl*, clang::FunctionProtoType const*, llvm::ArrayRef<clang::Expr*>, clang::SourceLocation, bool) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x6046103)
#24 0x00000000063ab0d6 clang::Sema::BuildCallToMemberFunction(clang::Scope*, clang::Expr*, clang::SourceLocation, llvm::MutableArrayRef<clang::Expr*>, clang::SourceLocation, clang::Expr*, bool, bool) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x63ab0d6)
#25 0x0000000006048386 clang::Sema::ActOnCallExpr(clang::Scope*, clang::Expr*, clang::SourceLocation, llvm::MutableArrayRef<clang::Expr*>, clang::SourceLocation, clang::Expr*) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x6048386)
#26 0x0000000005bdd442 clang::Parser::ParsePostfixExpressionSuffix(clang::ActionResult<clang::Expr*, true>) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x5bdd442)
#27 0x0000000005bde3d5 clang::Parser::ParseCastExpression(clang::Parser::CastParseKind, bool, bool&, clang::Parser::TypeCastState, bool, bool*) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x5bde3d5)
#28 0x0000000005bda4d9 clang::Parser::ParseAssignmentExpression(clang::Parser::TypeCastState) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x5bda4d9)
#29 0x0000000005bda3ce clang::Parser::ParseExpression(clang::Parser::TypeCastState) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x5bda3ce)
#30 0x0000000005c678d8 clang::Parser::ParseExprStatement(clang::Parser::ParsedStmtContext) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x5c678d8)
#31 0x0000000005c659d7 clang::Parser::ParseStatementOrDeclarationAfterAttributes(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*, clang::ParsedAttributes&, clang::ParsedAttributes&) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x5c659d7)
#32 0x0000000005c65223 clang::Parser::ParseStatementOrDeclaration(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x5c65223)
#33 0x0000000005c6ec60 clang::Parser::ParseCompoundStatementBody(bool) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x5c6ec60)
#34 0x0000000005c6fbf1 clang::Parser::ParseFunctionStatementBody(clang::Decl*, clang::Parser::ParseScope&) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x5c6fbf1)
#35 0x0000000005bb8dbf clang::Parser::ParseFunctionDefinition(clang::ParsingDeclarator&, clang::Parser::ParsedTemplateInfo const&, clang::Parser::LateParsedAttrList*) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x5bb8dbf)
#36 0x0000000005c25426 clang::Parser::ParseDeclGroup(clang::ParsingDeclSpec&, clang::DeclaratorContext, clang::ParsedAttributes&, clang::SourceLocation*, clang::Parser::ForRangeInit*) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x5c25426)
#37 0x0000000005bb742d clang::Parser::ParseDeclOrFunctionDefInternal(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec&, clang::AccessSpecifier) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x5bb742d)
#38 0x0000000005bb6d27 clang::Parser::ParseDeclarationOrFunctionDefinition(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*, clang::AccessSpecifier) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x5bb6d27)
#39 0x0000000005bb5bce clang::Parser::ParseExternalDeclaration(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x5bb5bce)
#40 0x0000000005bb35b0 clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, clang::Sema::ModuleImportState&) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x5bb35b0)
#41 0x0000000005bada8e clang::ParseAST(clang::Sema&, bool, bool) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x5bada8e)
#42 0x0000000004d68943 clang::FrontendAction::Execute() (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x4d68943)
#43 0x0000000004cec6cf clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x4cec6cf)
#44 0x0000000004e3b4a8 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x4e3b4a8)
#45 0x00000000028d279d cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x28d279d)
#46 0x00000000028ce615 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) driver.cpp:0:0
#47 0x00000000028cd4f9 clang_main(int, char**, llvm::ToolContext const&) (/home/dim/ins/llvmorg-17-init-08743-g6b8d19d2b57/bin/clang+0x28cd4f9)
```

Similarly, with the oldest (llvmorg-10-init-00000-g 8f5b44aead8):


```
Assertion failed: (getKind() == Template && "Unexpected kind"), function getAsTemplate, file /home/dim/src/llvm/llvm-project/clang/include/clang/AST/TemplateBase.h, line 281.
Stack dump:
0.      Program arguments: /home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang -cc1 -triple x86_64-- -S pr62331-1.cpp
1.      pr62331-1.cpp:18:14: current parser token ')'
2.      pr62331-1.cpp:16:13: parsing function body 'bug1'
3.      pr62331-1.cpp:16:13: in compound statement ('{}')
 #0 0x0000000001c6c978 llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x1c6c978)
 #1 0x0000000001c6a978 llvm::sys::RunSignalHandlers() (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x1c6a978)
 #2 0x0000000001c6d1f6 SignalHandler(int) Signals.cpp:0:0
 #3 0x0000000827ba9a60 handle_signal /usr/src/lib/libthr/thread/thr_sig.c:0:3
 #4 0x0000000827ba901f thr_sighandler /usr/src/lib/libthr/thread/thr_sig.c:247:1
 #5 0x00000008258b38a3 ([vdso]+0x2d3)
 #6 0x000000082b15f92a __sys_thr_kill /usr/obj/usr/src/lib/libc/thr_kill.S:4:0
 #7 0x000000082b0d89c4 __raise /usr/src/lib/libc/gen/raise.c:0:10
 #8 0x000000082b189779 abort /usr/src/lib/libc/stdlib/abort.c:73:17
 #9 0x000000082b0bb7a1 (/lib/libc.so.7+0x937a1)
#10 0x00000000032b2764 (anonymous namespace)::TemplateInstantiator::TransformTemplateName(clang::CXXScopeSpec&, clang::TemplateName, clang::SourceLocation, clang::QualType, clang::NamedDecl*, bool) SemaTemplateInstantiate.cpp:0:0
#11 0x00000000032a34f2 clang::TreeTransform<(anonymous namespace)::TemplateInstantiator>::TransformType(clang::TypeLocBuilder&, clang::TypeLoc) SemaTemplateInstantiate.cpp:0:0
#12 0x00000000032a26c8 clang::TreeTransform<(anonymous namespace)::TemplateInstantiator>::TransformType(clang::TypeSourceInfo*) SemaTemplateInstantiate.cpp:0:0
#13 0x00000000032b7f50 clang::TreeTransform<(anonymous namespace)::TemplateInstantiator>::TransformCXXUnresolvedConstructExpr(clang::CXXUnresolvedConstructExpr*) SemaTemplateInstantiate.cpp:0:0
#14 0x00000000032acec6 clang::Sema::SubstInitializer(clang::Expr*, clang::MultiLevelTemplateArgumentList const&, bool) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x32acec6)
#15 0x0000000002f521c2 clang::Sema::CheckCXXDefaultArgExpr(clang::SourceLocation, clang::FunctionDecl*, clang::ParmVarDecl*) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x2f521c2)
#16 0x0000000002f54748 clang::Sema::GatherArgumentsForCall(clang::SourceLocation, clang::FunctionDecl*, clang::FunctionProtoType const*, unsigned int, llvm::ArrayRef<clang::Expr*>, llvm::SmallVectorImpl<clang::Expr*>&, clang::Sema::VariadicCallType, bool, bool) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x2f54748)
#17 0x0000000002f52dac clang::Sema::ConvertArgumentsForCall(clang::CallExpr*, clang::Expr*, clang::FunctionDecl*, clang::FunctionProtoType const*, llvm::ArrayRef<clang::Expr*>, clang::SourceLocation, bool) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x2f52dac)
#18 0x000000000317dba7 clang::Sema::BuildCallToMemberFunction(clang::Scope*, clang::Expr*, clang::SourceLocation, llvm::MutableArrayRef<clang::Expr*>, clang::SourceLocation) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x317dba7)
#19 0x0000000002f3cec7 clang::Sema::BuildCallExpr(clang::Scope*, clang::Expr*, clang::SourceLocation, llvm::MutableArrayRef<clang::Expr*>, clang::SourceLocation, clang::Expr*, bool) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x2f3cec7)
#20 0x0000000002f55df9 clang::Sema::ActOnCallExpr(clang::Scope*, clang::Expr*, clang::SourceLocation, llvm::MutableArrayRef<clang::Expr*>, clang::SourceLocation, clang::Expr*) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x2f55df9)
#21 0x0000000002c084e1 clang::Parser::ParsePostfixExpressionSuffix(clang::ActionResult<clang::Expr*, true>) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x2c084e1)
#22 0x0000000002c0a876 clang::Parser::ParseCastExpression(bool, bool, bool&, clang::Parser::TypeCastState, bool) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x2c0a876)
#23 0x0000000002c05ce6 clang::Parser::ParseAssignmentExpression(clang::Parser::TypeCastState) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x2c05ce6)
#24 0x0000000002c05c1e clang::Parser::ParseExpression(clang::Parser::TypeCastState) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x2c05c1e)
#25 0x0000000002c5406a clang::Parser::ParseExprStatement(clang::Parser::ParsedStmtContext) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x2c5406a)
#26 0x0000000002c52ae1 clang::Parser::ParseStatementOrDeclarationAfterAttributes(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*, clang::Parser::ParsedAttributesWithRange&) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x2c52ae1)
#27 0x0000000002c52583 clang::Parser::ParseStatementOrDeclaration(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x2c52583)
#28 0x0000000002c5a510 clang::Parser::ParseCompoundStatementBody(bool) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x2c5a510)
#29 0x0000000002c5b094 clang::Parser::ParseFunctionStatementBody(clang::Decl*, clang::Parser::ParseScope&) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x2c5b094)
#30 0x0000000002bc33c0 clang::Parser::ParseFunctionDefinition(clang::ParsingDeclarator&, clang::Parser::ParsedTemplateInfo const&, clang::Parser::LateParsedAttrList*) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x2bc33c0)
#31 0x0000000002bdcfcc clang::Parser::ParseDeclGroup(clang::ParsingDeclSpec&, clang::DeclaratorContext, clang::SourceLocation*, clang::Parser::ForRangeInit*) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x2bdcfcc)
#32 0x0000000002bc27cd clang::Parser::ParseDeclOrFunctionDefInternal(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec&, clang::AccessSpecifier) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x2bc27cd)
#33 0x0000000002bc22b2 clang::Parser::ParseDeclarationOrFunctionDefinition(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec*, clang::AccessSpecifier) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x2bc22b2)
#34 0x0000000002bc10e9 clang::Parser::ParseExternalDeclaration(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec*) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x2bc10e9)
#35 0x0000000002bbf1e8 clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, bool) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x2bbf1e8)
#36 0x0000000002bba2cd clang::ParseAST(clang::Sema&, bool, bool) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x2bba2cd)
#37 0x00000000022c6283 clang::FrontendAction::Execute() (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x22c6283)
#38 0x0000000002282972 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x2282972)
#39 0x0000000002358ec3 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x2358ec3)
#40 0x0000000000847910 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/home/dim/ins/llvmorg-10-init-00000-g8f5b44aead8/bin/clang+0x847910)
```

Some people that recently (well, relatively :) touched `TemplateInstantiator::TransformTemplateName` include @zygoloid , @maskray, @mizvekov  and @kazutakahirata . Who has any clue?

**Comment by shafik** [comment]
Much reduced: https://godbolt.org/z/ePvKWj9KW

```cpp
template <class T1>
struct A {
  template <typename T2>
  void b(T2 a = T2{});
};

template <>
template <typename T2>
void A<void>::b(T2);

void f() {
    A<void>{}.b<int>();
}
```

**Comment by shafik** [comment]
I think we are getting the parameter wrong at some point b/c the following works but should be ill-formed:

```cpp
template <class T1>
struct A {
  template <typename T2>
  void b(T2 a = T1{});
};

template <>
template <typename T2>
void A<void>::b(T2);

void f() {
    A<void>{}.b<int>();
}
```

when I debug the original the case just above and break at the assertion I see the following:

```console
> expr ((TemplateTypeParmType*)((*Arg)->getType()).getTypePtr())->getDecl()->dump()
TemplateTypeParmDecl 0x13d04b820 <codegenCrash.cpp:3:13, col:22> col:22 referenced typename depth 1 index 0 T2
> expr ((SubstTemplateTypeParmType*)Ty.getNonReferenceType().getTypePtr())->getReplacedParameter()->dump()
TemplateTypeParmDecl 0x13d06a198 <codegenCrash.cpp:3:13, col:22> col:22 typename depth 0 index 0 T2
```

which does not look right. This feels vaguely familiar from other bugs that @erichkeane and @alexander-shaposhnikov are looking at

**Comment by erichkeane** [comment]
Alexander and I aren't really poking around these sorts of things, most of ours are concept related.  Its likely that we're getting the template instantiation args wrong (perhaps for the default arg?).  We MIGHT actually find with the out of line function template decl that `https://reviews.llvm.org/D146178` accidentially fixes it (by doing the right thing), but I'm not positive/won't hold that up on this.

**Comment by shafik** [comment]
Fixed in trunk: https://godbolt.org/z/Yn386qjn1

**Comment by DimitryAndric** [comment]
For people that look at this bug later on, this was fixed with `llvmorg-19-init-01176-g`752c172bc7d ("[Clang][Sema] fix outline member function template with default align crash (#80288)") by @jcsxky.
